core/amber/src/main/python/core/runnables/data_processor.py

```diff
@@ -29,9 +29,38 @@ def run(self) -> None:
         self._running.set()
         self._switch_context()
         while self._running.is_set():
+            self.process_state()
             self.process_tuple()
             self._switch_context()
 
+    def process_state(self) -> None:
+        state_ = self._context.tuple_processing_manager.get_input_state()
+        if state_ is not None:
+            try:
+                executor = self._context.executor_manager.executor
+                port_id = self._context.tuple_processing_manager.current_input_port_id
+                port: int
+                if port_id is None:
+                    # no upstream, special case for source executor.
+                    port = 0
+                else:
+                    port = port_id.id
+
+                with replace_print(
+                        self._context.worker_id,
+                        self._context.console_message_manager.print_buf,
+                ):
```

Comment:
wait.. you are duplicating the process tuple here. a few questions:
1. does port matter when processing states? (the answer might be yes, but please explain).
2. will there be print messages during processing of a state? I think no? the processing of a state is rather simple to me: it takes the key value paire and assign it to the executor. that's it.

Reply:
1. Port matters. process_state function can get different state from different input port. 
2. I think it's helpful to allow users to print the state in the console to help them debug.